base/multidimensional.jl

```diff
@@ -226,6 +226,14 @@ index_shape_dim(A, dim, ::Colon) = (trailingsize(A, dim),)
 @inline index_shape_dim(A, dim, i::AbstractArray{Bool}, I...) = (sum(i), index_shape_dim(A, dim+1, I...)...)
 @inline index_shape_dim{N}(A, dim, i::AbstractArray{CartesianIndex{N}}, I...) = (size(i)..., index_shape_dim(A, dim+N, I...)...)
 
+@inline decolon(A::AbstractVector, ::Colon) = (indices(A,1),)
+@inline decolon(A::AbstractArray, ::Colon) = (1:length(A),)
+@inline decolon(A::AbstractArray, I...) = decolon_dim(A, 1, I...)
+@inline decolon_dim(A::AbstractArray, dim)                = ()
+@inline decolon_dim(A::AbstractArray, dim, ::Colon)       = dim == ndims(A) ?  (indices(A, dim),) : (1:trailingsize(A, dim),)
+@inline decolon_dim(A::AbstractArray, dim, i1, I...)      = (i1, decolon_dim(A, dim+1, I...)...)
+@inline decolon_dim(A::AbstractArray, dim, ::Colon, I...) = (indices(A, dim), decolon_dim(A, dim+1, I...)...)
```

Comment:
It's a great observation, and I agree 100% with the sentiment. The problem is that `to_index` is intended for a single dimension and doesn't currently take the reference array as an input. This is more like `to_indexes`, which already takes a tuple of indices, but NOT the reference array. Since indices can be arrays, adding the array as a first argument seems likely to sow a lot of confusion, esp. for any non-base callers.